checker/src/call_graph.rs

```diff
@@ -0,0 +1,887 @@
+            included_crates: Vec::<Box<str>>::new(),
+        }
+    }
+}
+
+/// The type of a call graph node.
+#[derive(Debug, Clone, PartialEq)]
+enum NodeType {
+    /// Regular root
+    Root,
+    /// Crate root: Starting point for analysis (pub fn)
+    CRoot,
+}
+
+/// Nodes in the call graph are functions defined by the program
+/// that is being analyzed.
+#[derive(Debug, Clone)]
+struct CallGraphNode {
+    /// The name of the function (derived from its DefId).
+    name: Box<str>,
+    /// The type of the node.
+    ntype: NodeType,
+}
+
+impl CallGraphNode {
+    pub fn new_croot(defid: DefId) -> CallGraphNode {
+        CallGraphNode {
+            name: CallGraphNode::format_name(defid),
+            ntype: NodeType::CRoot,
+        }
+    }
+
+    pub fn new_root(defid: DefId) -> CallGraphNode {
+        CallGraphNode {
+            name: CallGraphNode::format_name(defid),
+            ntype: NodeType::Root,
+        }
+    }
+
+    /// Extracts a function name from the DefId of a function.
+    fn format_name(defid: DefId) -> Box<str> {
+        let tmp1 = format!("{:?}", defid);
+        let tmp2: &str = tmp1.split("~ ").collect::<Vec<&str>>()[1];
+        tmp2.replace(")", "").into_boxed_str()
+    }
+
+    /// A node is excluded if its name does not include any
+    /// one of the included crates' names.
+    pub fn is_excluded(&self, included_crates: &[&str]) -> bool {
+        let mut excluded = true;
+        for crate_name in included_crates.iter() {
+            if self.name.contains(crate_name) {
+                excluded = false;
+                break;
+            }
+        }
+        excluded
+    }
+
+    pub fn is_croot(&self) -> bool {
+        self.ntype == NodeType::CRoot
+    }
+}
+
+/// Denotes ways that two types, t1 and t2, can be related.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+enum TypeRelationKind {
+    /// States that t1 is equivalent to t2.
+    Eq,
+    /// States that t2 is a member of t1.
+    Member,
+}
+
+/// A type relation relates types `rtype1` and `rtype2` by a
+/// `TypeRelationKind`.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+struct TypeRelation {
+    kind: TypeRelationKind,
+    rtype1: Box<str>,
+    rtype2: Box<str>,
+}
+
+impl TypeRelation {
+    fn new_eq(t1: Box<str>, t2: Box<str>) -> TypeRelation {
+        TypeRelation {
+            kind: TypeRelationKind::Eq,
+            rtype1: t1,
+            rtype2: t2,
+        }
+    }
+
+    fn new_member(t1: Box<str>, t2: Box<str>) -> TypeRelation {
+        TypeRelation {
+            kind: TypeRelationKind::Member,
+            rtype1: t1,
+            rtype2: t2,
+        }
+    }
+}
+
+/// Types in the call graph as encoded as `RType` structs.
+/// An `RType` has:
+/// - `id`: A unique identifier.
+/// - `name`: A textual representation.
+/// - `type_relations`: A set of derived type relations.
+#[derive(Debug, Clone)]
+struct RType {
```

Comment:
RustType?

Reply:
Adding "Rust" seems redundant.  Perhaps "CallGraphType"?

Reply:
The types are specifically the types of callee's arguments, hence Rust-level program types rather than a type associated with the call graph itself. Perhaps I can call it "EdgeType"?